client/src/stream.rs

```diff
@@ -83,34 +93,65 @@ impl rpc::Server for CallbackServer {
                     output_frame_size,
                 );
 
+                let input_nbytes = nframes as usize * input_frame_size;
+                let output_nbytes = nframes as usize * output_frame_size;
+
                 // Clone values that need to be moved into the cpu pool thread.
-                let input_shm = unsafe { self.input_shm.as_ref().map(|shm| shm.unsafe_view()) };
-                let output_shm = unsafe { self.output_shm.as_ref().map(|shm| shm.unsafe_view()) };
+                let mut shm = unsafe { self.shm.as_ref().unwrap().unsafe_view() };
+
+                let duplex_copy_ptr = match &mut self.duplex_input {
+                    Some(buf) => {
+                        assert_eq!(self.dir, StreamDirection::Duplex);
+                        assert!(input_frame_size > 0);
+                        assert!(buf.capacity() >= input_nbytes);
+                        buf.as_mut_ptr()
+                    }
+                    None => ptr::null_mut(),
+                } as usize;
                 let user_ptr = self.user_ptr;
                 let cb = self.data_cb.unwrap();
+                let dir = self.dir;
 
                 self.cpu_pool.spawn_fn(move || {
-                    let input_ptr = match input_shm {
-                        Some(shm) => unsafe {
-                            shm.get_slice(nframes as usize * input_frame_size)
-                                .unwrap()
-                                .as_ptr()
+                    // Input and output reuse the same shmem backing.  Unfortunately, cubeb's data_callback isn't
+                    // specified in such a way that would require the callee to consume all of the input before
+                    // writing to the output (i.e., it is passed as two pointers that aren't expected to alias).
+                    // That means we need to copy the input here.
```

Comment:
Your understanding is correct in the first diagram.  Splitting the shmem is a trade-off, it requires 2x the size but avoids a copy - since we're currently over-estimating the required size (see my reply to your next question), a copy seems like the cheaper option than doubling the shmem requirement for each stream.  We can revisit once we have precisely sized shmems; there are other improvements that can be made in this area once that's addressed.